### ISSUE TITLE
New rule proposal: Flag duplicate literals

### ISSUE THREAD

**Issue by platinumazure** [eslint/eslint#7759]
**Please describe what the rule should do:**

This rule should flag uses of duplicate literals in the same file. The goal of this rule is to recommend that those duplicate literals be combined into one variable, which increases maintainability (by solving the problem of a user changing only a subset of the literals in the file, resulting in inconsistent behavior).

It is assumed that this will only work on literals in the same file. No attempt will be made to cross-reference other JS files to look for duplicates across files.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[ ] Warns about a potential error
[x] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
/**** 1. If conditions written as quick one-liners ****/
if (/myPattern/.test(str)) {
}

// Later...

if (/myPattern/.test(str) && somethingElse()) {
}

/**** 2. Validation configuration in popular libraries ****/

const MyModel = Backbone.Model.extend({
    validation: {
        WorkEmail: {
            pattern: /^[^<>:"^/|%?*\\]*$/,
            message: "Email cannot contain special characters"
        },
        PersonalEmail: {
            pattern: /^[^<>:"^/|%?*\\]*$/,                     // <-- Duplicated RegExp literal
            message: "Email cannot contain special characters" // <-- Duplicated string literal
        }
    }
});

/**** 3. Reusing a literal already in a variable (see Bikeshedding Question 1 below) ****/
const myRegEx = /^somePattern$/;

// Later...

if (/^somePattern$/.test(str)) {
    // ...
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This is a common maintainability refactor, library-independent, and can seriously save time when those duplicated literals need to be changed later.

---------------

Some bikeshedding questions:

1. Should this rule also flag a literal that is already in a variable (and simply reused), or should that fall under another rule? (See example 3 above)
1. What literals should and shouldn't be flagged, and what sort of options should we supply?
    * My proposal is we should flag string, RegExp, and numeric literals by default, and allow users to opt out of any of those types using basic rule options. It would also be great to allow specific exceptions.
    * I don't think we should flag boolean literals because they are ubiquitous and obvious and creating a variable just for a boolean is pretty silly most of the time. I wouldn't be opposed to adding an option to flag booleans (as an opt-in stricter check), but I don't think it's needed.
    * It might make sense to check only strings and RegExp literals at first, since there could be conflicts with no-magic-numbers for numeric literals. I'd be okay with that too.

I'll champion this.

**Comment by not-an-aardvark** [comment]
Tentatively :+1:, assuming we figure out a good solution to the concerns below.

> Should this rule also flag a literal that is already in a variable (and simply reused), or should that fall under another rule? (See example 3 above)

I think that should be flagged by this rule. In example 3, the user probably meant to use `myRegEx.test(str)`, and warning about that seems to be well within the scope of this rule.

> What literals should and shouldn't be flagged, and what sort of options should we supply?

I agree that flagging boolean literals (and `null`) by default might be overkill. There a few other literals that I don't think it makes sense to flag:

* `0`
* `1`
* `""`

---

I'm concerned about the rule flagging number literals in variable assignments. For example, I wouldn't want it to flag code like this:

```js
const DEFAULT_FOO_AMOUNT = 10;
const DEFAULT_BAR_AMOUNT = 10;
// (happens to be the same as DEFAULT_FOO_AMOUNT, but might change in the future)
```

On the other hand, I agree that it would be nice for the rule to flag code like this:

```js
const FOO_AMOUNT = 7361231019245809;

// (much later in the file)

// oops, maybe this person didn't realized that the variable is declared above
const AMOUNT_OF_FOO = 7361231019245809;
```

I feel like the fundamental difference between these two cases is that in the first case, the literals happen to be the same due to "coincidence", so it doesn't make sense to flag them. On the other hand, in the second case the literal is "complex" enough that the two literals are probably referring to the same conceptual value and violating DRY. To me, it's not obvious how the rule would be able to distinguish cases like these reliably, so unless we can figure out a good way to do that, maybe the rule should ignore duplicated literals if both of the literals appear in variable declarations.

(Note that this is also loosely related to [`no-magic-numbers`](http://eslint.org/docs/rules/no-magic-numbers), although I don't think there's enough of a conflict to warrant concern.)

**Comment by platinumazure** [comment]
@not-an-aardvark Thanks for your feedback. If the rule (as a first pass) didn't flag numeric literals, is that worth considering? In my experience, there will be plenty of value in catching duplicate string literals and RegExp literals already.

**Comment by not-an-aardvark** [comment]
Sure, that seems reasonable given that we already have `no-magic-numbers` to handle related issues with string literals.

**Comment by platinumazure** [comment]
@eslint/eslint-team Could we do a quick evaluation of this? It's been open long enough that we can certainly close, but I'd be grateful for maybe 1-2 days more for evaluation and then we can close if it doesn't get enough support in that time. Thanks!

**Comment by mysticatea** [comment]
I'm sorry for overlooked.
Sounds reasonable to me. One concern, `RegExp` hhas state (`RegExp#lastIndex`) for each instance, so the merging of `RegExp` literals needs careful.

**Comment by ljharb** [comment]
Specifically I think it'll be stateful if it's a global or sticky regex (g or y flags)

**Comment by not-an-aardvark** [comment]
I think the rule should disallow duplicates even for stateful regexes. The point of the rule is to avoid code duplication, which is still a concern even if the fix isn't trivial.

For example, one possible fix could look like this:

```js
const x = /complicated regex/g;
const y = /complicated regex/g;

// the user should fix this to:

const x = /complicated regex/g;
const y = new RegExp(x);
```

**Comment by ljharb** [comment]
Why is that necessarily better? Regex literals always work, the global `RegExp` could have been modified.

**Comment by not-an-aardvark** [comment]
Because it reduces code duplication, which makes the code easier to maintain in the event that the regex needs to be changed. In most cases, a modified `RegExp` global is not a significant concern (if globals have been tampered with, a lot of bets are off anyway).

**Comment by ljharb** [comment]
Code intending to be robust (to operate in a world where all bets are off, such as all of my shim modules) would use a regex literal to remove that concern. The fact is that it's not *safe* to replace a regex literal with a non-literal, so I don't think it's a good idea for this rule to do it. Code duplication is only a problem when it's unnecessary, but it might be for regex literals.

**Comment by not-an-aardvark** [comment]
Fair enough, but I think needing to account for untrustworthy globals is fairly uncommon. So in my opinion, it's still worthwhile for the rule to report duplicate regexes. If you need to avoid using `RegExp` because all global bets are off, that seems like a case where it would be appropriate to use a disable comment.

> The fact is that it's not *safe* to replace a regex literal with a non-literal, so I don't think it's a good idea for this rule to do it.

Just to make sure we're on the same page here -- I wasn't suggesting that the rule should automatically fix your code, just that it should flag duplicates because there is a better way to write the code in most cases.

**Comment by ljharb** [comment]
Fair point.

It might make sense then for this rule to have individual options for each kind of literal - that way I can disable it for regexes, for example, without losing the benefit of all the other kinds?

**Comment by alberto** [comment]
@ljharb what would be the other kinds be? Strings?

Wouldn't it make sense to create a specific rule for each type? `no-magic-literals` looks like a good parallel for `no-magic-numbers`.

**Comment by not-an-aardvark** [comment]
I'm not necessarily opposed to creating a specific rule for each type, but the rule in this proposal is different from `no-magic-numbers`. `no-magic-numbers` flags *any* number literals (with a few exceptions) outside of variable declarations, whereas this rule would flag *duplicate* literals even if they appear in variable declarations.

**Comment by alberto** [comment]
We could add an option to enforce that, too.

**Comment by not-an-aardvark** [comment]
Could you explain the value of doing that? I don't understand why we would want to add options to make this similar `no-magic-numbers`. In my mind, the rules are enforcing different things -- `no-magic-numbers` requires "magic" values to be given a semantic meaning, whereas `no-duplicate-literals` prevents code duplication. I agree that there's some overlap there, but the rules have significantly different goals, so I'm not convinced it's worth trying to combine them. I think having a `no-magic-numbers` and a `no-duplicate-literals` rule would be fine.

**Comment by platinumazure** [comment]
Strongly agree with @not-an-aardvark that this proposal is completely different from no-magic-numbers.

**Comment by ljharb** [comment]
(+1 for different rules)

The options I would envision would allow separately configured enforcement of duplicate literals of: numbers, strings, booleans, regexes, and perhaps arrays containing other literals, and objects containing other literals.

**Comment by alberto** [comment]
Just thinking out loud, no strong opinions here, just a diffuse idea.

I see this rule targeting strings, regexes and maybe numbers. Not so much for the other cases mentioned.

It looks like the would be enforceable per type (i.e. disable it for regexes or numbers). It also looks like there would be exceptions at least for numbers and strings. So I started to think maybe they would be better as different rules, even if the implementations are very similar (maybe we could deduplicate that under the hood).

Also, I think the exceptions (which are configurable) would have to be the same between `no-magic-numbers` and this rule. Also would apply to strings if we created a separate `no-magic-strings`. (I can't think of an exceptions for a regex I would like to ignore).

And I still see value in a rule flagging regexes even if they are not duplicated, to avoid inlining them like @platinumazure example 1:
```
/^[^<>:"^/|%?*\\]*$/.test(str)
```

So it occurred to me that maybe it should be one rule per type? I see an similar underlying concern I'm not sure how to verbalize.

**Comment by michaelficarra** [comment]
1. `no-magic-numbers` is definitely unrelated.
1. Separate rules for separate types makes sense.
1. The rule for RegExps is a bad idea. It advocates for a potentially hazardous change. RegExps are objects. They have mutable state and object identity. Numbers and strings don't have these issues. I would strongly vote against inclusion of a rule like this. It is the equivalent to asking someone to dedupe object initialisers, array initialisers, or functions just because they're defined using the same source text.
1. The numeric literal deduping should probably account for immediate negation of a numeric literal as a separate "literal". People probably wouldn't want `5` and `-5` being treated as dupes.

Overall, I don't think this rule is that helpful. I may have a program where I do

```js
const HTTP_PORT = 80;
const MAX_CONCURRENT_CONNECTIONS = 80;
```

I would argue that is much better than what this rule would suggest I do

```js
const HTTP_PORT = 80;
const MAX_CONCURRENT_CONNECTIONS = HTTP_PORT; // what?!
```

**Comment by mikesherov** [comment]
Seems like the true use case here is "no duplicate regex literals outside of assignment expression". That is to say, numeric and string literals with equivalent values assigned to separate bindings makes sense at least some of the time when two constants just might have same value, but:

```
if (/a/.test(b)) {

}

if (/a/.test(c)) {

}
```

Can always be refactored safely, no? It only becomes unsafe when stateful methods of a regex are used...

**Comment by michaelficarra** [comment]
Or object identity (`/a/ === /a/`). And it's not just assignment, but any time they escape. This includes when they are passed as an argument as well as when they are the `this` value in a method call.

**Comment by mikesherov** [comment]
Sure, said another way, really the only time we want to do this is when they are used as part of a callExpression memberExpression, right?

**Comment by michaelficarra** [comment]
Not really. I was saying that even `/a/.test(...)` leaks that regex as the `this` value in whatever's currently assigned to `RegExp.prototype.test`. But it's a fair compromise to allow that. I just don't think the rule is a good idea in general.

**Comment by platinumazure** [comment]
@eslint/eslint-team We're pretty close to accepting this, but we should close this if we can't resolve the open questions.

Open questions as I see them:

1. Objects we want to cover: Strings (I think yes), regexes (some concerns raised about state implications), numbers (probably not)?
1. Separate rules: I'm okay with that as long as the team doesn't feel there's a maintainability overhead, anyone else strongly feel one way or the other?

My own thoughts re: regex literals: To me this is one of those "slightly more opinionated" rules where it intentionally emphasizes one coding virtue even in cases where it might be wrong due to circumstance (in this case, the stateful nature of the regexes). In those cases, disable comments can always be used. If it seems likely that users will be *commonly* using multiple regex literals that actually need to have separate state, so that they're writing disable comments a lot, then the rule shouldn't look at regexes. But my feeling is that the need for *identical* literals *with independent state* is pretty rare and so disable comments would be sufficient in those cases.

**Comment by not-an-aardvark** [comment]
@platinumazure Are you still championing this?

**Comment by platinumazure** [comment]
@not-an-aardvark Thanks, I really shouldn't, I'd forgotten about this and definitely wouldn't have time to implement. If I get time later or if someone volunteers to run with it, I may create a new issue.

Let's close for now.
